src/components/ProjectSearch/index.js

```diff
@@ -20,84 +23,256 @@ import {
+      0
+    );
+  };
+
+  onKeyDown = (e: SyntheticKeyboardEvent<HTMLInputElement>) => {
+    if (e.key === "Escape") {
+      return;
+    }
+
+    e.stopPropagation();
+
+    if (e.key !== "Enter") {
+      return;
+    }
+    this.focusedItem = null;
+    this.props.searchSources(this.state.inputValue);
+  };
+
+  onEnterPress = () => {
+    if (this.focusedItem && !this.state.inputFocused) {
+      const { setExpanded, file, expanded, match } = this.focusedItem;
+      if (setExpanded) {
+        setExpanded(file, !expanded);
+      } else if (match) {
+        this.selectMatchItem(match);
+      }
+    }
+  };
+
+  inputOnChange = (e: SyntheticInputEvent<HTMLInputElement>) => {
+    const inputValue = e.target.value;
+    this.setState({ inputValue });
+  };
+
+  renderFile(
+    file: Result,
+    focused: boolean,
+    expanded: boolean,
+    setExpanded: Function
+  ) {
+    if (focused) {
+      this.focusedItem = { setExpanded, file, expanded };
+    }
+
+    const matchesLength = file.matches.length;
+    const matches = ` (${matchesLength} match${matchesLength > 1 ? "es" : ""})`;
+
+    return (
+      <div
+        className={classnames("file-result", { focused })}
+        key={file.sourceId}
+        onClick={e => setExpanded(file, !expanded)}
+      >
+        <Svg name="arrow" className={classnames({ expanded })} />
+        <img className="file" />
+        <span className="file-path">{getRelativePath(file.filepath)}</span>
+        <span className="matches-summary">{matches}</span>
+      </div>
+    );
   }
 
-  isProjectSearchEnabled() {
-    return this.props.activeSearch === "project";
+  renderMatch(match: Match, focused: boolean) {
+    if (focused) {
+      this.focusedItem = { match };
+    }
+    return (
+      <div
+        className={classnames("result", { focused })}
+        onClick={() => setTimeout(() => this.selectMatchItem(match), 50)}
+      >
+        <span className="line-number" key={match.line}>
+          {match.line}
+        </span>
+        {this.renderMatchValue(match)}
+      </div>
+    );
   }
 
-  renderTextSearch() {
-    const {
-      sources,
-      results,
-      status,
-      searchSources,
-      closeProjectSearch,
-      selectLocation,
-      textSearchQuery
-    } = this.props;
+  renderMatchValue(lineMatch: Match) {
+    return highlightMatches(lineMatch);
+  }
+
+  renderResults() {
+    const results = this.getResults().filter(
+      result => result.matches.length > 0
+    );
+
+    const { status } = this.props;
+
+    function getFilePath(item: Match, index?: number) {
+      return item.filepath
+        ? `${item.sourceId}-${index || "$"}`
+        : `${item.sourceId}-${item.line}-${item.column}-${index || "$"}`;
+    }
+
+    const renderItem = (item, depth, focused, _, expanded, { setExpanded }) => {
+      return item.filepath
+        ? this.renderFile(item, focused, expanded, setExpanded)
+        : this.renderMatch(item, focused);
+    };
```

Comment:
why define `renderItem` in the scope and not on the class?

Reply:
I had lifted this from the previous implementation in most cases so some of it wasn't scrutinized much. Good question though!

Reply:
Moving the `renderItem` body up to the class does require adding a function signature to make Flow happy too. So that may have been why it was originally put within.